redis/maintenance_events.py

```diff
@@ -0,0 +1,496 @@
+        """
+        return hash((self.__class__, self.id))
+
+
+class MaintenanceEventsConfig:
+    """
+    Configuration class for maintenance events handling behaviour. Events are received through
+    push notifications.
+
+    This class defines how the Redis client should react to different push notifications
+    such as node moving, migrations, etc. in a Redis cluster.
+
+    """
+
+    def __init__(
+        self,
+        enabled: bool = False,
+        proactive_reconnect: bool = True,
+        relax_timeout: Optional[Number] = 20,
+    ):
+        """
+        Initialize a new MaintenanceEventsConfig.
+
+        Args:
+            enabled (bool): Whether to enable maintenance events handling.
+                Defaults to False.
+            proactive_reconnect (bool): Whether to proactively reconnect when a node is replaced.
+                Defaults to True.
+            relax_timeout (Number): The relax timeout to use for the connection during maintenance.
+                If -1 is provided - the relax timeout is disabled. Defaults to 20.
+
+        """
+        self.enabled = enabled
+        self.relax_timeout = relax_timeout
+        self.proactive_reconnect = proactive_reconnect
+
+    def __repr__(self) -> str:
+        return (
+            f"{self.__class__.__name__}("
+            f"enabled={self.enabled}, "
+            f"proactive_reconnect={self.proactive_reconnect}, "
+            f"relax_timeout={self.relax_timeout}, "
+            f")"
+        )
+
+    def is_relax_timeouts_enabled(self) -> bool:
+        """
+        Check if the relax_timeout is enabled. The '-1' value is used to disable the relax_timeout.
+        If relax_timeout is set to None, it will make the operation blocking
+        and waiting until any response is received.
+
+        Returns:
+            True if the relax_timeout is enabled, False otherwise.
+        """
+        return self.relax_timeout != -1
+
+
+class MaintenanceEventPoolHandler:
+    def __init__(
+        self,
+        pool: Union["ConnectionPool", "BlockingConnectionPool"],
+        config: MaintenanceEventsConfig,
+    ) -> None:
+        self.pool = pool
+        self.config = config
+        self._processed_events = set()
+        self._lock = threading.RLock()
+        self.connection = None
+
+    def set_connection(self, connection: "ConnectionInterface"):
+        self.connection = connection
+
+    def remove_expired_notifications(self):
+        with self._lock:
+            for notification in tuple(self._processed_events):
+                if notification.is_expired():
+                    self._processed_events.remove(notification)
+
+    def handle_event(self, notification: MaintenanceEvent):
+        self.remove_expired_notifications()
+
+        if isinstance(notification, NodeMovingEvent):
+            return self.handle_node_moving_event(notification)
+        else:
+            logging.error(f"Unhandled notification type: {notification}")
```

Comment:
If we don't expect to handle any other events then `NodeMovingEvent` why does signature allows it's parent class to be passed (`MaintenanceEvent`)